File: curriculum/algos/RadixSort.java
final class RadixSort {
static void radixSort(int[] values) {
int[] buffer = new int[values.length];
for (int shift = 0; shift < 32; shift += 8) {
int[] counts = new int[256];
for (int value : values) ++counts[(value >>> shift) & 255];
for (int index = 1; index < counts.length; ++index) counts[index] += counts[index - 1];
for (int index = values.length - 1; index >= 0; --index) {
int value = values[index];
int digit = (value >>> shift) & 255;
buffer[--counts[digit]] = value;
}
int[] temporary = values;
values = buffer;
buffer = temporary;
}
}
}

File: curriculum/algos/radix_sort.rs
pub fn radix_sort(values: &mut [u32]) {
let mut buffer = values.to_vec();
for shift in (0..32).step_by(8) {
let mut counts = [0usize; 256];
for &value in values.iter() { counts[((value >> shift) & 255) as usize] += 1; }
let mut offsets = [0usize; 256];
for index in 1..256 { offsets[index] = offsets[index - 1] + counts[index - 1]; }
for &value in values.iter() {
let digit = ((value >> shift) & 255) as usize;
buffer[offsets[digit]] = value;
offsets[digit] += 1;
}
values.copy_from_slice(&buffer);
}
}
